File: jar/src/main/java/jline/api/mam/m3pp/M3pp_interleave_fitc_trace.java
/**
 * @file M3PP interleaved fitting from empirical trace data
 *
 * Implements M3PP interleaving and fitting directly from empirical trace data
 * with multiple arrival classes.
 *
 * @since LINE 3.0
 */
package jline.api.mam.m3pp;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import jline.util.Pair;
import jline.util.matrix.MatrixCell;

public final class M3pp_interleave_fitc_trace {
    private M3pp_interleave_fitc_trace() {}

    public static Pair<MatrixCell, List<MatrixCell>> m3pp_interleave_fitc_trace(
            double[] interArrivalTimes, int[] classLabels) {
        return m3pp_interleave_fitc_trace(interArrivalTimes, classLabels, null, null, null);
    }

    public static Pair<MatrixCell, List<MatrixCell>> m3pp_interleave_fitc_trace(
            double[] interArrivalTimes, int[] classLabels, Double t, Double tinf) {
        return m3pp_interleave_fitc_trace(interArrivalTimes, classLabels, t, tinf, null);
    }

    /**
     * Interleaves k M3PP to fit a multi-class trace with m classes.
     */
    public static Pair<MatrixCell, List<MatrixCell>> m3pp_interleave_fitc_trace(
            double[] interArrivalTimes, int[] classLabels,
            Double t, Double tinf, boolean[][] mapping) {
        if (interArrivalTimes.length != classLabels.length) {
            throw new IllegalArgumentException(
                    "Inter-arrival times and class labels must have the same length");
        }

        TreeSet<Integer> uniqueLabelsSet = new TreeSet<Integer>();
        for (int label : classLabels) uniqueLabelsSet.add(label);
        Integer[] uniqueLabels = uniqueLabelsSet.toArray(new Integer[0]);
        int m = uniqueLabels.length;

        double meanInterArrival = average(interArrivalTimes);
        double totalTime = sum(interArrivalTimes);
        double computedT = (t != null) ? t : (10.0 * meanInterArrival);
        double computedTinf = (tinf != null) ? tinf : Math.max(10.0 * computedT, totalTime / 100.0);

        double[][] countsT = computeMulticlassCountsAtScale(interArrivalTimes, classLabels, computedT);
        double[][] countsTinf = computeMulticlassCountsAtScale(interArrivalTimes, classLabels, computedTinf);

        double arrivalRate = 1.0 / meanInterArrival;

        double[] classRates = new double[m];
        for (int i = 0; i < m; i++) {
            int classCount = 0;
            for (int label : classLabels) {
                if (label == uniqueLabels[i]) classCount++;
            }
            classRates[i] = ((double) classCount / classLabels.length) * arrivalRate;
        }

        boolean[][] finalMapping = (mapping != null) ? mapping : computeCorrelationBasedMapping(countsTinf, m, 0.75);
        int k = finalMapping[0].length;

        validateMapping(finalMapping, m);

        System.out.println("Fitting " + m + " classes with " + k + " M3PP(2,m_j) processes");

        boolean[][] filters = new boolean[k][m];
        for (int j = 0; j < k; j++) {
            for (int i = 0; i < m; i++) {
                filters[j][i] = finalMapping[i][j];
            }
        }

        double[] processRates = new double[k];
        for (int j = 0; j < k; j++) {
            double sum = 0.0;
            for (int i = 0; i < m; i++) {
                if (filters[j][i]) sum += classRates[i];
            }
            processRates[j] = sum;
        }

        double[][] classRatesPerProcess = new double[k][];
        for (int j = 0; j < k; j++) {
            List<Double> arr = new ArrayList<Double>();
            for (int i = 0; i < m; i++) {
                if (filters[j][i]) arr.add(classRates[i]);
            }
            classRatesPerProcess[j] = new double[arr.size()];
            for (int i = 0; i < arr.size(); i++) classRatesPerProcess[j][i] = arr.get(i);
        }

        double[] idcT = new double[k];
        double[] idcTinf = new double[k];

        for (int j = 0; j < k; j++) {
            double[] aggregateCountsT = new double[countsT.length];
            for (int wIdx = 0; wIdx < countsT.length; wIdx++) {
                double s = 0.0;
                for (int classIdx = 0; classIdx < m; classIdx++) {
                    if (filters[j][classIdx]) s += countsT[wIdx][classIdx];
                }
                aggregateCountsT[wIdx] = s;
            }

            double[] aggregateCountsTinf = new double[countsTinf.length];
            for (int wIdx = 0; wIdx < countsTinf.length; wIdx++) {
                double s = 0.0;
                for (int classIdx = 0; classIdx < m; classIdx++) {
                    if (filters[j][classIdx]) s += countsTinf[wIdx][classIdx];
                }
                aggregateCountsTinf[wIdx] = s;
            }

            double meanT = average(aggregateCountsT);
            double varT = 0.0;
            for (double v : aggregateCountsT) varT += (v - meanT) * (v - meanT);
            varT /= aggregateCountsT.length;
            idcT[j] = varT / (processRates[j] * computedT);

            double meanTinf = average(aggregateCountsTinf);
            double varTinf = 0.0;
            for (double v : aggregateCountsTinf) varTinf += (v - meanTinf) * (v - meanTinf);
            varTinf /= aggregateCountsTinf.length;
            idcTinf[j] = varTinf / (processRates[j] * computedTinf);
        }

        double[][] gtc = new double[k][];
        for (int j = 0; j < k; j++) {
            int numClassesInProcess = 0;
            for (boolean f : filters[j]) if (f) numClassesInProcess++;
            double[] result = new double[numClassesInProcess];

            int resultIdx = 0;
            for (int i = 0; i < m; i++) {
                if (filters[j][i]) {
                    double[] classCounts = new double[countsT.length];
                    for (int wIdx = 0; wIdx < countsT.length; wIdx++) {
                        classCounts[wIdx] = countsT[wIdx][i];
                    }
                    double classMean = average(classCounts);
                    double classVariance = 0.0;
                    for (double v : classCounts) classVariance += (v - classMean) * (v - classMean);
                    classVariance /= classCounts.length;

                    double[] otherClassesCounts = new double[countsT.length];
                    for (int wIdx = 0; wIdx < countsT.length; wIdx++) {
                        double s = 0.0;
                        for (int classIdx = 0; classIdx < m; classIdx++) {
                            if (filters[j][classIdx] && classIdx != i) {
                                s += countsT[wIdx][classIdx];
                            }
                        }
                        otherClassesCounts[wIdx] = s;
                    }

                    double covariance = 0.0;
                    boolean any = false;
                    for (double v : otherClassesCounts) if (v != 0.0) { any = true; break; }
                    if (any) {
                        double otherMean = average(otherClassesCounts);
                        for (int idx = 0; idx < classCounts.length; idx++) {
                            covariance += (classCounts[idx] - classMean) * (otherClassesCounts[idx] - otherMean);
                        }
                        covariance /= classCounts.length;
                    }

                    result[resultIdx] = classVariance + covariance;
                    resultIdx++;
                }
            }
            gtc[j] = result;
        }

        return M3pp_interleave_fitc.m3pp_interleave_fitc(processRates, idcT, idcTinf,
                classRatesPerProcess, gtc, computedT, computedTinf, null, false);
    }

    private static double[][] computeMulticlassCountsAtScale(
            double[] interArrivalTimes, int[] classLabels, double scale) {
        TreeSet<Integer> uniqueLabelsSet = new TreeSet<Integer>();
        for (int label : classLabels) uniqueLabelsSet.add(label);
        Integer[] uniqueLabels = uniqueLabelsSet.toArray(new Integer[0]);
        int m = uniqueLabels.length;
        double totalTime = sum(interArrivalTimes);
        int numWindows = Math.max(1, (int) (totalTime / scale));

        double[][] counts = new double[numWindows][m];

        double currentTime = 0.0;
        int windowIndex = 0;

        for (int i = 0; i < interArrivalTimes.length; i++) {
            currentTime += interArrivalTimes[i];

            double targetWindowEnd = (windowIndex + 1) * scale;
            int classIdx = indexOf(uniqueLabels, classLabels[i]);
            if (currentTime <= targetWindowEnd) {
                if (classIdx >= 0) {
                    counts[windowIndex][classIdx] += 1.0;
                }
            } else {
                windowIndex++;
                if (windowIndex >= numWindows) break;
                if (classIdx >= 0) {
                    counts[windowIndex][classIdx] = 1.0;
                }
            }
        }
        return counts;
    }

    private static int indexOf(Integer[] arr, int v) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i].intValue() == v) return i;
        }
        return -1;
    }

    private static boolean[][] computeCorrelationBasedMapping(double[][] counts, int m, double threshold) {
        double[][] covariance = computeCovarianceMatrix(counts);
        double[] variance = new double[m];
        for (int i = 0; i < m; i++) {
            double[] classCounts = new double[counts.length];
            for (int j = 0; j < counts.length; j++) classCounts[j] = counts[j][i];
            double mean = average(classCounts);
            double v = 0.0;
            for (double cc : classCounts) v += (cc - mean) * (cc - mean);
            variance[i] = v / classCounts.length;
        }

        Set<Integer> pool = new HashSet<Integer>();
        for (int i = 0; i < m; i++) pool.add(i);
        List<List<Integer>> groups = new ArrayList<List<Integer>>();

        while (!pool.isEmpty()) {
            int pivot = -1;
            double maxVar = Double.NEGATIVE_INFINITY;
            for (int i : pool) {
                if (variance[i] > maxVar) {
                    maxVar = variance[i];
                    pivot = i;
                }
            }
            if (pivot < 0) break;

            List<Integer> currentGroup = new ArrayList<Integer>();
            currentGroup.add(pivot);
            pool.remove(pivot);

            Set<Integer> toRemove = new HashSet<Integer>();
            for (int h : pool) {
                double correlation = (variance[pivot] > 0 && variance[h] > 0)
                        ? covariance[pivot][h] / Math.sqrt(variance[pivot] * variance[h])
                        : 0.0;
                if (correlation >= threshold) {
                    currentGroup.add(h);
                    toRemove.add(h);
                }
            }
            pool.removeAll(toRemove);
            groups.add(currentGroup);
        }

        int k = groups.size();
        boolean[][] mapping = new boolean[m][k];
        for (int j = 0; j < k; j++) {
            for (int classIdx : groups.get(j)) {
                mapping[classIdx][j] = true;
            }
        }
        return mapping;
    }

    private static double[][] computeCovarianceMatrix(double[][] counts) {
        int m = counts[0].length;
        double[] means = new double[m];
        for (int i = 0; i < m; i++) {
            double s = 0.0;
            for (int j = 0; j < counts.length; j++) s += counts[j][i];
            means[i] = s / counts.length;
        }

        double[][] covariance = new double[m][m];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < m; j++) {
                double s = 0.0;
                for (int kk = 0; kk < counts.length; kk++) {
                    s += (counts[kk][i] - means[i]) * (counts[kk][j] - means[j]);
                }
                covariance[i][j] = s / counts.length;
            }
        }
        return covariance;
    }

    private static void validateMapping(boolean[][] mapping, int m) {
        if (mapping.length != m) {
            throw new IllegalArgumentException("Number of classes does not match mapping");
        }
        for (int i = 0; i < m; i++) {
            int count = 0;
            for (boolean b : mapping[i]) if (b) count++;
            if (count != 1) {
                throw new IllegalArgumentException(
                        "Invalid mapping: class " + i + " mapped to " + count + " processes");
            }
        }
    }

    private static double average(double[] arr) {
        double s = 0.0;
        for (double v : arr) s += v;
        return s / arr.length;
    }

    private static double sum(double[] arr) {
        double s = 0.0;
        for (double v : arr) s += v;
        return s;
    }
}

File: jar/src/main/kotlin/jline/api/mam/m3pp/M3pp_interleave_fitc_trace.kt
/**
 * @file M3PP interleaved fitting from empirical trace data
 * 
 * Implements M3PP interleaving and fitting directly from empirical trace data 
 * with multiple arrival classes. Extracts statistical characteristics from 
 * inter-arrival time data for data-driven parameter estimation.
 * 
 * @since LINE 3.0
 */
package jline.api.mam.m3pp

import jline.util.matrix.MatrixCell
import jline.util.matrix.Matrix
import kotlin.math.*

/**
 * Interleaves k M3PP to fit a multi-class trace with m classes.
 * 
 * @param interArrivalTimes Array of inter-arrival times
 * @param classLabels Array of class labels corresponding to each arrival
 * @param t Finite time scale (optional, computed from trace if not provided)
 * @param tinf Near-infinite time scale (optional, computed from trace if not provided)
 * @param mapping m x k binary matrix mapping the m classes to k m3pp (optional, computed if not provided)
 * @return Pair of (fitted MMAP, list of component M3PPs)
 */
fun m3pp_interleave_fitc_trace(
    interArrivalTimes: DoubleArray,
    classLabels: IntArray,
    t: Double? = null,
    tinf: Double? = null,
    mapping: Array<BooleanArray>? = null
): Pair<MatrixCell, List<MatrixCell>> {
    
    require(interArrivalTimes.size == classLabels.size) {
        "Inter-arrival times and class labels must have the same length"
    }
    
    // Number of classes
    val uniqueLabels = classLabels.distinct().sorted()
    val m = uniqueLabels.size
    
    // Compute time scales if not provided
    val meanInterArrival = interArrivalTimes.average()
    val totalTime = interArrivalTimes.sum()
    val computedT = t ?: (10.0 * meanInterArrival)
    val computedTinf = tinf ?: maxOf(10.0 * computedT, totalTime / 100.0)
    
    // Compute counting processes
    val countsT = computeMulticlassCountsAtScale(interArrivalTimes, classLabels, computedT)
    val countsTinf = computeMulticlassCountsAtScale(interArrivalTimes, classLabels, computedTinf)
    
    // Overall arrival rate
    val arrivalRate = 1.0 / meanInterArrival
    
    // Per-class rates
    val classRates = DoubleArray(m) { i ->
        val classCount = classLabels.count { it == uniqueLabels[i] }
        (classCount.toDouble() / classLabels.size) * arrivalRate
    }
    
    // Compute or create mapping
    val finalMapping = mapping ?: computeCorrelationBasedMapping(countsTinf, m)
    val k = finalMapping[0].size // Number of M3PP processes
    
    // Validate mapping
    validateMapping(finalMapping, m)
    
    println("Fitting $m classes with $k M3PP(2,m_j) processes")
    
    // Create filters for each M3PP
    val filters = Array(k) { j ->
        BooleanArray(m) { i -> finalMapping[i][j] }
    }
    
    // Total rates for each M3PP
    val processRates = DoubleArray(k) { j ->
        classRates.filterIndexed { i, _ -> filters[j][i] }.sum()
    }
    
    // Per-class rates within each M3PP
    val classRatesPerProcess = Array(k) { j ->
        classRates.filterIndexed { i, _ -> filters[j][i] }.toDoubleArray()
    }
    
    // Compute IDC for each M3PP
    val idcT = DoubleArray(k)
    val idcTinf = DoubleArray(k)
    
    for (j in 0 until k) {
        // Aggregate counts for j-th M3PP
        val aggregateCountsT = DoubleArray(countsT.size) { windowIdx ->
            uniqueLabels.indices
                .filter { filters[j][it] }
                .sumOf { classIdx -> countsT[windowIdx][classIdx] }
        }
        
        val aggregateCountsTinf = DoubleArray(countsTinf.size) { windowIdx ->
            uniqueLabels.indices
                .filter { filters[j][it] }
                .sumOf { classIdx -> countsTinf[windowIdx][classIdx] }
        }
        
        // Compute IDC
        val meanT = aggregateCountsT.average()
        val varT = aggregateCountsT.map { (it - meanT) * (it - meanT) }.average()
        idcT[j] = varT / (processRates[j] * computedT)
        
        val meanTinf = aggregateCountsTinf.average()
        val varTinf = aggregateCountsTinf.map { (it - meanTinf) * (it - meanTinf) }.average()
        idcTinf[j] = varTinf / (processRates[j] * computedTinf)
    }
    
    // Compute per-class variance plus marginal covariance for each M3PP
    val gtc = Array(k) { j ->
        val numClassesInProcess = filters[j].count { it }
        val result = DoubleArray(numClassesInProcess)
        
        var resultIdx = 0
        for (i in 0 until m) {
            if (filters[j][i]) {
                // Per-class variance
                val classCounts = DoubleArray(countsT.size) { windowIdx -> countsT[windowIdx][i] }
                val classMean = classCounts.average()
                val classVariance = classCounts.map { (it - classMean) * (it - classMean) }.average()
                
                // Marginal covariance with other classes in the same M3PP
                val otherClassesCounts = DoubleArray(countsT.size) { windowIdx ->
                    uniqueLabels.indices
                        .filter { classIdx -> filters[j][classIdx] && classIdx != i }
                        .sumOf { classIdx -> countsT[windowIdx][classIdx] }
                }
                
                val covariance = if (otherClassesCounts.any { it != 0.0 }) {
                    val otherMean = otherClassesCounts.average()
                    classCounts.indices.map { idx ->
                        (classCounts[idx] - classMean) * (otherClassesCounts[idx] - otherMean)
                    }.average()
                } else {
                    0.0
                }
                
                result[resultIdx] = classVariance + covariance
                resultIdx++
            }
        }
        
        result
    }
    
    // Call the existing interleave fitting function
    return m3pp_interleave_fitc(
        av = processRates,
        btv = idcT,
        binfv = idcTinf,
        acc = classRatesPerProcess,
        gtcc = gtc,
        t = computedT,
        tinf = computedTinf,
        mapping = null
    )
}

/**
 * Compute multiclass counting processes at a given time scale
 */
private fun computeMulticlassCountsAtScale(
    interArrivalTimes: DoubleArray, 
    classLabels: IntArray, 
    scale: Double
): Array<DoubleArray> {
    
    val uniqueLabels = classLabels.distinct().sorted()
    val m = uniqueLabels.size
    val totalTime = interArrivalTimes.sum()
    val numWindows = maxOf(1, (totalTime / scale).toInt())
    
    val counts = Array(numWindows) { DoubleArray(m) }
    
    var currentTime = 0.0
    var windowIndex = 0
    
    for (i in interArrivalTimes.indices) {
        currentTime += interArrivalTimes[i]
        
        val targetWindowEnd = (windowIndex + 1) * scale
        if (currentTime <= targetWindowEnd) {
            val classIdx = uniqueLabels.indexOf(classLabels[i])
            if (classIdx >= 0) {
                counts[windowIndex][classIdx] += 1.0
            }
        } else {
            // Move to next window
            windowIndex++
            if (windowIndex >= numWindows) break
            
            val classIdx = uniqueLabels.indexOf(classLabels[i])
            if (classIdx >= 0) {
                counts[windowIndex][classIdx] = 1.0
            }
        }
    }
    
    return counts
}

/**
 * Compute correlation-based mapping using covariance analysis
 */
private fun computeCorrelationBasedMapping(
    counts: Array<DoubleArray>,
    m: Int,
    threshold: Double = 0.75
): Array<BooleanArray> {
    
    // Compute covariance matrix
    val covariance = computeCovarianceMatrix(counts)
    val variance = DoubleArray(m) { i ->
        val classCounts = DoubleArray(counts.size) { j -> counts[j][i] }
        val mean = classCounts.average()
        classCounts.map { (it - mean) * (it - mean) }.average()
    }
    
    // Group classes based on correlation
    val pool = (0 until m).toMutableSet()
    val groups = mutableListOf<List<Int>>()
    
    while (pool.isNotEmpty()) {
        // Find class with highest variance among remaining
        val pivot = pool.maxByOrNull { variance[it] } ?: break
        val currentGroup = mutableListOf(pivot)
        pool.remove(pivot)
        
        // Find correlated classes
        val toRemove = mutableSetOf<Int>()
        for (h in pool) {
            val correlation = if (variance[pivot] > 0 && variance[h] > 0) {
                covariance[pivot][h] / sqrt(variance[pivot] * variance[h])
            } else {
                0.0
            }
            
            if (correlation >= threshold) {
                currentGroup.add(h)
                toRemove.add(h)
            }
        }
        
        pool.removeAll(toRemove)
        groups.add(currentGroup)
    }
    
    // Create mapping matrix
    val k = groups.size
    val mapping = Array(m) { BooleanArray(k) }
    
    for (j in groups.indices) {
        for (classIdx in groups[j]) {
            mapping[classIdx][j] = true
        }
    }
    
    return mapping
}

/**
 * Compute covariance matrix for count data
 */
private fun computeCovarianceMatrix(counts: Array<DoubleArray>): Array<DoubleArray> {
    val m = counts[0].size
    val means = DoubleArray(m) { i ->
        counts.map { it[i] }.average()
    }
    
    val covariance = Array(m) { DoubleArray(m) }
    
    for (i in 0 until m) {
        for (j in 0 until m) {
            covariance[i][j] = counts.indices.map { k ->
                (counts[k][i] - means[i]) * (counts[k][j] - means[j])
            }.average()
        }
    }
    
    return covariance
}

/**
 * Validate mapping matrix
 */
private fun validateMapping(mapping: Array<BooleanArray>, m: Int) {
    require(mapping.size == m) { "Number of classes does not match mapping" }
    
    // Check that each class is mapped to exactly one M3PP
    for (i in 0 until m) {
        val mappingCount = mapping[i].count { it }
        require(mappingCount == 1) { "Invalid mapping: class $i mapped to $mappingCount processes" }
    }
}
/**
 * M3Pp Interleave Fit Count Trace algorithms
 */
@Suppress("unused")
class M3ppInterleaveFitCountTraceAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}